### PR TITLE
feat: Publish org stats to kafka

### PR DESCRIPTION
**Pull request by alex-hofsteede**

### CODE REVIEW

**Comment by alex-hofsteede** [comment]
@JTCunning this is just here to start a conversation. There are a couple of outstanding things to consider.

1. There are some cases like duplicate events, too large events etc. that we currently raise `APIForbidden` for but we don't track counts for, perhaps we should include these with `reason` in the new snuba metrics.

2. For CORS errors, we track them as `tsdb.models.project_total_received_cors` but _not_ against `tsdb.models.(project/org/key)_total_received` I think it wold be nicer/cleaner if everything got counted towards `total_received` and the cors stuff was just another outcome/reason, but that would technically change the way the the counters work currently.

File: src/sentry/web/api.py [hunk 1]
```diff
@@ -94,35 +101,54 @@ def wrapped(request, *args, **kwargs):
...
+    if outcome == 'filtered':
+        increment_list.extend([
+            (tsdb.models.project_total_blacklisted, project_id),
+            (tsdb.models.organization_total_blacklisted, org_id),
+            (tsdb.models.key_total_blacklisted, key_id),
+        ])
+        if reason in FILTER_STAT_KEYS_TO_VALUES:
+            increment_list.append((FILTER_STAT_KEYS_TO_VALUES[reason], project_id))
+
+    elif outcome == 'rate_limited':
```
Comment (alex-hofsteede) [inline src/sentry/web/api.py]: unclear if outcome=filtered and outcome=rate_limited should both just be outcome=rejected (ie outcome is boolean) with a different `reason` for each

Comment (alex-hofsteede) [reply src/sentry/web/api.py]: that way outcome would be binary

File: src/sentry/web/api.py [hunk 2]
```diff
@@ -94,35 +101,70 @@ def wrapped(request, *args, **kwargs):
...
+            (tsdb.models.key_total_rejected, key_id),
+        ])
+
+    tsdb.incr_multi(
+        increment_list,
+        timestamp=to_datetime(time()),
+    )
+
+    # Send a snuba metrics payload.
+    kafka_outcomes_publisher.publish(
```
Comment (JTCunning) [inline src/sentry/web/api.py]: Let's put this behind a sample rate option.

File: src/sentry/web/api.py [hunk 3]
```diff
@@ -145,19 +187,7 @@ def process_event(event_manager, project, key, remote_addr, helper, attachments)
...
-                (tsdb.models.organization_total_received,
-                    project.organization_id),
-                (tsdb.models.organization_total_rejected,
-                    project.organization_id),
-                (tsdb.models.key_total_received, key.id),
-                (tsdb.models.key_total_rejected, key.id),
-            ],
-            timestamp=tsdb_start_time,
-        )
+        track_outcome(project.organization_id, project.id, key.id, 'rate_limited', None)
```
Comment (JTCunning) [inline src/sentry/web/api.py]: I feel like theres a rate_limit reason that we track somewhere based on which quota returned `is_limited`.

File: src/sentry/web/api.py [hunk 4]
```diff
@@ -909,7 +931,7 @@ def post(self, request, project, helper, **kwargs):
         origin = instance.get_origin()
         if not is_valid_origin(origin, project):
             if project:
-                tsdb.incr(tsdb.models.project_total_received_cors, project.id)
+                track_outcome(project.organization_id, project.id, None, 'invalid', 'cors')
```
Comment (JTCunning) [inline src/sentry/web/api.py]: Does this not have a request key associated with the event?

File: src/sentry/web/api.py [hunk 5]
```diff
@@ -407,8 +429,7 @@ def _dispatch(self, request, helper, project_id=None, origin=None, *args, **kwar
             if not project:
                 raise APIError('Client must be upgraded for CORS support')
             if not is_valid_origin(origin, project):
-                tsdb.incr(tsdb.models.project_total_received_cors,
-                          project.id)
+                track_outcome(project.organization_id, project.id, None, 'invalid', 'cors')
```
Comment (JTCunning) [inline src/sentry/web/api.py]: Does this not have a request key associated with the event?

File: src/sentry/web/api.py (hunk 1 above)
Comment (JTCunning) [reply src/sentry/web/api.py]: Product-wise, I think it makes sense to split them. All of our graphs are going to have to roll up based on these outcomes, and the hovercard details would split by reason, so it's easier for me to see queries that group by outcome and reason, and the UI possibly flattening them.

**Review by JTCunning — COMMENTED** [review]
Leaving comments for now.

Logic looks good to me, but I'd be more comfortable with another engineer approving such a sensitive codepath.

File: src/sentry/web/api.py (hunk 5 above)
Comment (alex-hofsteede) [reply src/sentry/web/api.py]: yes, but we fetch it a few lines down inside an `if`, and I'm not 100% confident I can rearrange these things without subtly breaking something.

File: src/sentry/web/api.py (hunk 1 above)
Comment (alex-hofsteede) [reply src/sentry/web/api.py]: yeah I think adding outcome=invalid also led me to this conclusion.

**Comment by alex-hofsteede** [comment]
There are currently a bunch of other places in here (Bad security reports, invalid minidumps, etc.) that currently raise `APIError` but have no outcome tracked. We can either go through here and add this tracking to all possible outcomes for a super rich dataset, or we can deliberately leave that out of scope for now.

File: src/sentry/web/api.py (hunk 5 above)
Comment (JTCunning) [reply src/sentry/web/api.py]: I guess it does save the trip to verifying a Key. We can nix it for now and always add it back in later.

**Review by mitsuhiko — APPROVED** [review]

**Pull request merged**